source/docs/async_http_filters.md

```diff
@@ -0,0 +1,46 @@
+## How to write http filters with asynchronous features
+
+To get started on writing a basic http filter, see [http-filter-example](https://github.com/envoyproxy/envoy-filter-example/tree/main/http-filter-example).
+
+This documentation is to explain the less obvious behaviors when a filter performs some asynchronous action, such as performing something in another thread, making a grpc call, etc.
+
+### How to wait for an asynchronous callback
+
+If you send a callback and your filter needs the response before it can complete the current step (e.g. `decodeHeaders`, `encodeHeaders`, etc.) the function must return one of the states that indicates work should not continue, e.g. [`Http::FilterHeadersStatus::StopIteration`](https://github.com/envoyproxy/envoy/blob/2d82c10a467cbb933ed5cb9bdb7eaae4ffa160de/envoy/http/filter.h#L45). For more details on this, see [flow_control.md](flow_control.md).
```

Comment:
Replace: `you send a callback and your filter needs the response` with `your extension initiates a long running operation and needs to wait for its completion`. I think the `send callback` is a bit unclear on what is actually hapenning.